[HOLD] [WIP] Introduce "Mark as unread" functionality in Expensify.cash

Gonals opened the pull request:
cc @roryabraham 

HELD on https://github.com/Expensify/Expensify.cash/pull/1977


### Details
The "New" marker has been modified so that it no longer disappears after a few seconds. It will behave like Slack's marker, which remains until the user leaves the chat. 

### Fixed Issues


### Tests
1. Log into e.cash and open a conversation with existing messages
2. On the hover menu of one of the messages, select "mark as unread":
<img width="1233" alt="Screen Shot 2021-03-16 at 3 43 38 PM" src="https://user-images.githubusercontent.com/19366280/111328469-8e7ccc00-866e-11eb-8f24-eed1d7d87d5b.png">
3. The `new` marker should show up over that message and the chat in the LHN should become bolded:
<img width="1602" alt="Screen Shot 2021-03-16 at 3 43 54 PM" src="https://user-images.githubusercontent.com/19366280/111328582-a6ece680-866e-11eb-90c7-b94ecb1e76b5.png">

5. Switch chats. It should remain bolded
6. Switch back to the original chat. The `new` marker should be correctly in place and the chat in the LHN should unbold in about 3 seconds. The `new` marker should remain in place after those 3 seconds.
7. Switch out and back in. the marker should be gone.
8. Do some regression testing to make sure the new marker shows correctly for actual new messages


### Tested On

- [x] Web
- [ ] Mobile Web
- [x] Desktop
- [ ] iOS
- [ ] Android

Gonals committed:
Handle Mark as Unread action

Gonals committed:
make the call correctly

Gonals committed:
Merge master

Gonals committed:
remove animation and refactor new marker

Gonals committed:
merge master

Gonals committed:
merge master

Gonals committed:
merge master

roryabraham commented on src/components/UnreadActionIndicator.js:
```diff
@@ -1,29 +1,17 @@
-UnreadActionIndicator.propTypes = propTypes;
 UnreadActionIndicator.displayName = 'UnreadActionIndicator';
 
 export default UnreadActionIndicator;
```
I think since this is now a plain component that just displays stuff statically, I think you should be able to wrap it in the `React.memo` HOC to prevent unnecessary re-renders.

roryabraham commented on src/libs/actions/Report.js:
```diff
@@ -204,13 +204,14 @@ function fetchChatReportsByIDs(chatList) {
+ * @param {Object} report
  */
-function setLocalLastRead(reportID, sequenceNumber) {
+function setLocalLastRead(reportID, sequenceNumber, report = null) {
```
Maybe it would be simpler to just pass the `maxSequenceNumber` rather than the whole report object?

roryabraham commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -25,6 +26,11 @@ const propTypes = {
     // Should the comment have the appearance of being grouped with the previous comment?
     displayAsGroup: PropTypes.bool.isRequired,
 
+    //
```
Missing comments for this and the next props

roryabraham reviewed (commented):
Overall, I think we could clean this up by leveraging `Onyx` a bit more to manage unreads and determine if the `UnreadActionIndicator` should show. This is what I'm thinking:

`onMarkAsUnread` in `ReportActionsView` is really just using the `updateLastReadActionID` Onyx action in `Reports.js`, then setting the `unreadActionCount` state. That state is ultimately only used to determine if a `ReportActionItem` should show the `UnreadActionIndicator` (something that's not even done in this component). What we could do instead is:

1) Add a `callback` field to each object in the `CONTEXT_ACTIONS` array in `ReportActionContextMenu`, and use that as the `onPress` handler in render. That eliminates the need for the extra prop, or to dynamically check the `text` or ID of the `contextAction` to determine the callback to use.
2) Make that callback just use the `updateLastReadActionID` Onyx action ... the `ReportActionContextMenu` already has all the info it needs in props to supply the parameters for that function.
3) You can then bind the `ReportActionItem` to the `report` it's attached to in Onyx and check the `unreadActionCount` field of that report to determine whether or not to show the `UnreadActionIndicator`

I think this will keep things a lot cleaner, especially once we have a half-dozen different context actions. What do you think?

Gonals commented:
> Overall, I think we could clean this up by leveraging `Onyx` a bit more to manage unreads and determine if the `UnreadActionIndicator` should show. This is what I'm thinking:
> 
> `onMarkAsUnread` in `ReportActionsView` is really just using the `updateLastReadActionID` Onyx action in `Reports.js`, then setting the `unreadActionCount` state. That state is ultimately only used to determine if a `ReportActionItem` should show the `UnreadActionIndicator` (something that's not even done in this component). What we could do instead is:
> 
> 1. Add a `callback` field to each object in the `CONTEXT_ACTIONS` array in `ReportActionContextMenu`, and use that as the `onPress` handler in render. That eliminates the need for the extra prop, or to dynamically check the `text` or ID of the `contextAction` to determine the callback to use.
> 2. Make that callback just use the `updateLastReadActionID` Onyx action ... the `ReportActionContextMenu` already has all the info it needs in props to supply the parameters for that function.
> 3. You can then bind the `ReportActionItem` to the `report` it's attached to in Onyx and check the `unreadActionCount` field of that report to determine whether or not to show the `UnreadActionIndicator`
> 
> I think this will keep things a lot cleaner, especially once we have a half-dozen different context actions. What do you think?

I... think I'm not a fan 😅 
1/2. While we could create all the callback functions in ReportActionContextMenu, that goes against our standard practices where we have the bulk of the logic in a larger component higher up and pass down whichever callbacks we need to (pinging @tgolen, since I had this discussion with him a while ago 😆)

3. Now that we have moved the marker inside `ReportActionItem`, we may be able to do something like this, although we'll still need to somehow make sure we are not displaying it for new messages that arrive while the chat is open, so we might still need the logic in `ReportActionsView` (not sure about this one. I'd have to play around for a bit, but that's my. gut feeling)

Gonals committed:
cleanup

Gonals committed:
cleanup

Gonals commented:
- - -
   ***I have read the CLA Document and I hereby sign the CLA***
   - - -

Gonals commented:
recheck

Gonals commented:
I have read the CLA Document and I hereby sign the CLA

Gonals commented:
Off WIP for now with a couple of caveats:
- If we decide to implement @roryabraham's suggested changes, it might go WIP again for a little bit
- I haven't been able to get the menu to show on mobile, so I haven't tested there yet

Gonals committed:
Make sure we don't move the New marker when new messages come in

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -204,13 +204,14 @@ function fetchChatReportsByIDs(chatList) {
     // Update the report optimistically
     Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT}${reportID}`, {
-        unreadActionCount: 0,
+        unreadActionCount: maxSequenceNumber > 0 ? maxSequenceNumber - sequenceNumber : 0,
```
- I'm not sure about why we need to pass the `maxSequenceNumber` in from the outside here. I think we can maybe see what the `maxSequenceNumber` is by looking somewhere in `actions/Report`?

- Using `Math.max()` might clean up the conditional.

- Also, slightly unsure about when the `sequenceNumber` would be different from the `maxSequenceNumber` probably it would be good to add a comment that explains this.

Gonals replied:
SequenceNumber will be different, for example, in this flow, when we mark an older sequence as unread

marcaaron replied:
Ok it looks better to me now and I think is self explanatory looking again. Thanks :)

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -663,14 +664,15 @@ function addAction(reportID, text, file) {
+function updateLastReadActionID(reportID, sequenceNumber, ignoreOrder = false) {
     const currentMaxSequenceNumber = reportMaxSequenceNumbers[reportID];
-    if (sequenceNumber < currentMaxSequenceNumber) {
+    if (!ignoreOrder && sequenceNumber < currentMaxSequenceNumber) {
```
Rather than add an `ignoreOrder` flag that basically says "don't run this function" would it make more sense to not call the function in the first place.

Gonals replied:
? It only skips checking the order, though. It runs the rest of the function

marcaaron replied:
Hmm maybe we should remove that logic as well.. this logic is confusing and even with the added comment I'm having some trouble understanding in what other contexts this should be used.

marcaaron replied:
Sorry, I'm trying to think of a better suggestion for this...

marcaaron replied:
Ok so here is my confusion and would like to get @tgolen's thoughts to see if I'm missing something.

I'm not sure I understand why we need to get the last action from the sorted actions here to mark everything as read...

https://github.com/Expensify/Expensify.cash/blob/c5cb5076dc092ee5246e3410e4557fc2d55b4a88/src/pages/home/report/ReportActionsView.js#L266-L278

Here we are looking at the `sequenceNumber` on the last action... but this information is already available to the `actions/Report` and `maxVisibleSequenceNumber` is kind of a misnomer since it has nothing to do with visibility.

We are filtering out the "loading" actions, but that too could be done in the `actions/Report` file I _think_.

If that's true then we wouldn't need an `ignoreOrder` and could make `sequenceNumber` optional and the logic for this method could then end up being

```js
function updateLastReadActionID(reportID, sequenceNumber) {
    const currentMaxSequenceNumber = reportMaxSequenceNumbers[reportID];
    const lastReadSequenceNumber = sequenceNumber 
        ? sequenceNumber 
        : currentMaxSequenceNumber;
    //... set the local last read 
}
```
Not sure if we need to do this change right now, but it seems more intuitive to "set to the most recent `sequenceNumber` or the one specified".

tgolen replied:
I want to dig into this a little so I'm going to pull the branch down and dive into it. I want to make sure I understand everything before offering up further comments on this thread. Thanks for being patient with me!

tgolen replied:
>  it seems more intuitive to "set to the most recent sequenceNumber when none is passed or use the one specified"

I love coming at it from this angle, and I agree it is much more intuitive.

> but that too could be done in the actions/Report file I think.

I briefly looked at this, and I am pretty concerned that changing `reportMaxSequenceNumbers` to not include loading actions has a high risk of breaking things. I would be more comfortable with keeping a separate `reportMaxSequenceNumbersForNonLoadingActions` (lol at that ugly name). 

It makes a lot of sense to me to move this logic from the view into the action file: https://github.com/Expensify/Expensify.cash/blob/c5cb5076dc092ee5246e3410e4557fc2d55b4a88/src/pages/home/report/ReportActionsView.js#L266-L278

Gonals replied:
I'm afraid I'm the one who is lost now.
What do you mean by loading/not loading actions?

tgolen replied:
In the `ReportActionsView.js`, when it calls `updateLastReadActionID`, it is passing the `maxVisibileSequenceNumber`. It gets that number by getting all the report actions and filtering out the ones with `action.loading`. This is what we are referring to.

What I would like to see for a solution is to have `ReportActionsView.js` updated so that all it does is this:
```
    recordMaxAction() {
        updateLastReadActionID(this.props.reportID);
    }
```

and also:
```
onMarkAsUnread={() => updateLastReadActionID(this.props.reportID, item.action.sequenceNumber - 1)}
``` 

All the rest of the logic will then be encapsulated into the `actions/Report.js` file.

Does that help clarify what we are asking for? It doesn't spell out the changes needed in `Report.js`, so let me know if you need help with that as well. I think this will be enough to point you in the right direction though.

marcaaron commented on src/libs/actions/Report.js:
```diff
@@ -663,14 +664,15 @@ function addAction(reportID, text, file) {
     }
 
-    setLocalLastRead(reportID, sequenceNumber);
+    setLocalLastRead(reportID, sequenceNumber, currentMaxSequenceNumber);
```
I don't think we need to add `currentMaxSequenceNumber` here. We're passing this to a method in the same scope that can access the same stuff.

marcaaron commented on src/pages/home/report/ReportActionContextMenu.js:
```diff
@@ -65,7 +68,7 @@ const defaultProps = {
 
 const ReportActionContextMenu = (props) => {
-    const wrapperStyle = getReportActionContextMenuStyles(props.isMini);
+    const wrapperStyle = getReportActionContextMenuStyles(true);
```
Was this just for testing or are we supposed to set this `true`?

Gonals replied:
Oh, nice catch!

marcaaron commented on src/pages/home/report/ReportActionContextMenu.js:
```diff
@@ -74,6 +77,7 @@ const ReportActionContextMenu = (props) => {
                     text={contextAction.text}
                     isMini={props.isMini}
                     key={contextAction.text}
+                    onPress={contextAction.text === 'Mark as Unread' ? props.onMarkAsUnread : () => {}}
```
`Mark as Unread` should be using a constant if we are going to string comparison check like this that way if someone changes the text they don't have to remember updating this here.

marcaaron commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -56,7 +63,8 @@ class ReportActionItem extends Component {
             || this.props.displayAsGroup !== nextProps.displayAsGroup
-            || !_.isEqual(this.props.action, nextProps.action);
+            || !_.isEqual(this.props.action, nextProps.action)
+            || this.props.displayNewIndicator !== nextProps.displayNewIndicator;
```
NAB, probably better to put this check before the deep equality check before it so that we avoid a more expensive calculation in favor of a cheaper one.

marcaaron commented on src/pages/home/report/ReportActionItem.js:
```diff
@@ -134,13 +144,15 @@ class ReportActionItem extends Component {
                                         isVisible
                                         reportID={-1}
                                         reportActionID={-1}
+                                        onMarkAsUnread={() => {}}
```
It would be better to default this to an empty function by using `defaultProps` of `ReportActionContextMenu`  vs explicitly passing an empty function.

marcaaron commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -39,6 +37,9 @@ const propTypes = {
         unreadActionCount: PropTypes.number,
+
+        // Number of the last sequence
+        maxSequenceNumber: PropTypes.number,
```
Can't we always figure this out if we are able to access `reportActions`?

Gonals replied:
Maybe with `sortedReportActions`? But I'm not really adding anything new here. The report is already being passed and it already has the `maxSequenceNumber` in it. React just complained if I accessed it without declaring it inside the shape, so it is fine this way, right?

marcaaron replied:
Looks unused now?

marcaaron commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -68,19 +70,18 @@ class ReportActionsView extends React.Component {
         this.shouldShowUnreadActionIndicator = true;
 
         this.state = {
+            unreadActionCount: 0,
```
I think this should be referring to the Onyx state and not local state. 🤔

marcaaron commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -113,6 +118,7 @@ class ReportActionsView extends React.Component {
         // We have switched to a new report
         if (prevProps.reportID !== this.props.reportID) {
             this.reset(prevProps.reportID);
+            this.shouldShowUnreadActionIndicator = true;
```
Was this a bug? I may have introduced this or broken this feature 🙈

Gonals replied:
Merging master did break stuff, but adding this line fixed it 😆

marcaaron commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -155,6 +161,13 @@ class ReportActionsView extends React.Component {
         }
     }
 
+    onMarkAsUnread(actionIndex) {
```
Needs method doc

marcaaron commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -155,6 +161,13 @@ class ReportActionsView extends React.Component {
+    onMarkAsUnread(actionIndex) {
+        updateLastReadActionID(this.props.reportID, actionIndex, true);
+        this.setState({
+            unreadActionCount: this.props.report.maxSequenceNumber - actionIndex,
```
The component should re-render with the correct `unreadActionCount` when we make the optimistic update?

Gonals replied:
Yup. It does so. That makes sense, right?

marcaaron replied:
Oh sorry I was saying we should not need to set the state to trigger the re-render but looks like you've got it now thanks.

marcaaron commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -165,17 +178,9 @@ class ReportActionsView extends React.Component {
-            }, 3000));
-        }
+        this.setState({
+            unreadActionCount: this.props.report.unreadActionCount,
```
Similar to other comments already left, but why can we not use the props directly?

marcaaron commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -335,15 +340,14 @@ class ReportActionsView extends React.Component {
                     displayAsGroup={this.isConsecutiveActionMadeByPreviousActor(index)}
                     onLayout={onLayout}
                     needsLayoutCalculation={needsLayoutCalculation}
+                    onMarkAsUnread={() => this.onMarkAsUnread(item.action.sequenceNumber - 1)}
```
It will be better for performance to pass the `sequenceNumber` from `ReportActionItem` and just pass `this.onMarkAsUnread` directly like:

```jsx
onMarkAsUnread={this.onMarkAsUnread}
```

as far as React is concerned seeing `() => someMethod()` means it gets interpreted as a "new function" on every render. This is usually OK but probably should be avoided in our most performance sensitive components like this one.

tgolen reviewed (changes requested):
> Add a callback field to each object in the CONTEXT_ACTIONS array in ReportActionContextMenu, and use that as the onPress handler in render. That eliminates the need for the extra prop, or to dynamically check the text or ID of the contextAction to determine the callback to use.

I very much agree with this suggestion. This is exactly what we are doing with the LHN and it has proven to be valuable. 

> While we could create all the callback functions in ReportActionContextMenu, that goes against our standard practices where we have the bulk of the logic in a larger component higher up and pass down whichever callbacks we need to 

I see this a little bit differently. I view Rory's suggestion as following exactly what you are saying the standard practice is. The higher-up component will be doing the logic (by virtue of the callback).

> dynamically check the text or ID of the contextAction to determine the callback to use

I think this is the part that stood out to me the most about this PR. It like an anti-pattern to have a dynamic callback as opposed to an explicit one. I think it also makes the component more difficult to use and implement because you need to have pretty in-depth knowledge of how the component works in order to implement callbacks for any of the other context menu items.

marcaaron commented:
As for the broader conversation above 

>You can then bind the ReportActionItem to the report it's attached to in Onyx and check the unreadActionCount field of that report to determine whether or not to show the UnreadActionIndicator

I think this is good advice. The other suggestion I am neutral on.

>especially once we have a half-dozen different context actions

Maybe it is better to wait for this scenario.

>I very much agree with this suggestion. This is exactly what we are doing with the LHN and it has proven to be valuable.

Full disclosure, I actually removed this logic because tightly coupling the a generic "option" with a hardcoded callback behavior made them harder to use in different contexts. In this case, I don't think it matters as much since these are less likely to be re-used so might not actually be important to change yet.

Gonals commented on src/components/UnreadActionIndicator.js:
```diff
@@ -1,29 +1,17 @@
 UnreadActionIndicator.displayName = 'UnreadActionIndicator';
 
-export default UnreadActionIndicator;
+export default memo(UnreadActionIndicator);
```
There's not a lot behind it. I read a bit about it and it seemed like a good idea, since we don't really. want to rerender the marker that often, right?

tgolen replied:
@Gonals for further reading, [this article](https://dmitripavlutin.com/use-react-memo-wisely/) really walks through the use of `memo` and when to use it.

In general, I think this is a good guideline:

> Use the following rule of thumb: don’t use memoization if you can’t quantify the performance gains.

Gonals replied:
Cool! Thanks guys! I'll remove it

Gonals commented on src/components/UnreadActionIndicator.js:
```diff
@@ -1,29 +1,17 @@
 UnreadActionIndicator.displayName = 'UnreadActionIndicator';
 
-export default UnreadActionIndicator;
+export default memo(UnreadActionIndicator);
```
I read a bit about it and it seemed like a good idea, since we don't really. want to rerender the marker that often, right?

Gonals commented on src/libs/actions/Report.js:
```diff
@@ -663,14 +664,15 @@ function addAction(reportID, text, file) {
  *
  * @param {Number} reportID
  * @param {Number} sequenceNumber
+ * @param {Boolean} ignoreOrder
```
Yup. Basically, it prevents the function from checking that the new "latest read" is at the end of the conversation. That check is relevant when new messages come in, but not when we are manually marking messages as unread.

Gonals committed:
Big revamp to remove some local variables and clean up code

Gonals commented:
Ok! I think I have addressed all comments.
I've changed both how we check the unreadActions (now we check the report from Onyx) and how the onMarAsRead callback works.
I think it is working fine, but I plan to do more testing tomorrow, since I get the feeling SOMETHING must be failing with how things are now (I haven't found what, yet).

Thanks, everyone!

marcaaron commented on src/pages/home/report/ReportActionContextMenu.js:
```diff
@@ -57,14 +22,57 @@ const propTypes = {
+    /**
+     * A list of all the context actions in this menu.
+     */
+    const CONTEXT_ACTIONS = [
```
NAB, one small side effect of this change is that we are now recreating this object with each render. I don't think it will cause problems tho.

roryabraham replied:
Rather than recreating this object with every render, I think we should convert this to a class component, make `contextActions` a property of the class, then set it once in `componentDidMount`

marcaaron replied:
Or the `constructor()`

roryabraham reviewed (changes requested):
Overall, I'm still not understanding the benefit to declaring the `onMarkAsRead` prop in `ReportActionsView` and drilling it down. All `onMarkAsRead` is doing is:

```javascript
() => updateLastReadActionID(this.props.reportID, item.action.sequenceNumber - 1, true)
```

So it's using an Onyx action with data that's all available in `ReportActionContextMenu`, where the callback could just be:

```javascript
onPress: () => updateLastReadActionID(this.props.reportID, this.props.reportActionID - 1, true)
```

That eliminates the extra clutter and unnecessary complexity of having an additional prop and drilling it down.

> that goes against our standard practices where we have the bulk of the logic in a larger component higher up and pass down whichever callbacks we need to

Overall I'm not really sure I see this pattern as beneficial in an environment where we have a global data store to control our UI. Since we've already agreed to manage the unread data in Onyx, then the manipulation of that data should be controlled where the manipulation is done. Why should the already very complex `ReportActionsView` be responsible for managing all the functionality of the `ReportActionContextMenu` as well, when it's just writing data to the global store anyways?

Also, it looks like you accidentally committed a log file.

Gonals committed:
remove memo and unused report

Gonals committed:
Merge branch 'master' into alberto-markUnread

Gonals commented:
> Overall, I'm still not understanding the benefit to declaring the `onMarkAsRead` prop in `ReportActionsView` and drilling it down. All `onMarkAsRead` is doing is:
> 
> ```js
> () => updateLastReadActionID(this.props.reportID, item.action.sequenceNumber - 1, true)
> ```
> 
> So it's using an Onyx action with data that's all available in `ReportActionContextMenu`, where the callback could just be:
> 
> ```js
> onPress: () => updateLastReadActionID(this.props.reportID, this.props.reportActionID - 1, true)
> ```
> 
> That eliminates the extra clutter and unnecessary complexity of having an additional prop and drilling it down.
> 
> > that goes against our standard practices where we have the bulk of the logic in a larger component higher up and pass down whichever callbacks we need to
> 
> Overall I'm not really sure I see this pattern as beneficial in an environment where we have a global data store to control our UI. Since we've already agreed to manage the unread data in Onyx, then the manipulation of that data should be controlled where the manipulation is done. Why should the already very complex `ReportActionsView` be responsible for managing all the functionality of the `ReportActionContextMenu` as well, when it's just writing data to the global store anyways?
> 
> Also, it looks like you accidentally committed a log file.

I'm partial to having the logic "higher up", especially since we'll be introducing other actions soon, but I don't really feel strongly about it. @marcaaron, @tgolen, any preferences on your side?

Gonals committed:
linter

Gonals committed:
remove log file

Gonals commented:
Sending this back to WIP as removing the local variable did break a few things as I expected. I'll try to fix it (or bring back the local variable if I can't). Let's keep the conversation above going in the meantime, though!

Gonals commented:
Hey people! Since we were having several discussions and I kept reworking this, I have decided to split it in two!
For now, I'm sending you all to https://github.com/Expensify/Expensify.cash/pull/1977 for the "New" marker fix. I'm still relying in a local variable, but I'm open to suggestions!

Gonals commented:
Moving to https://github.com/Expensify/Expensify.cash/pull/2433, since we kinda redid the whole thing!
